virtcontainers/device/api/interface.go

```diff
@@ -27,35 +27,43 @@ func DeviceLogger() *logrus.Entry {
 // DeviceReceiver is an interface used for accepting devices
 // a device should be attached/added/plugged to a DeviceReceiver
 type DeviceReceiver interface {
+	// these are for hotplug/hot-unplug devices to/from hypervisor
 	HotplugAddDevice(Device, config.DeviceType) error
 	HotplugRemoveDevice(Device, config.DeviceType) error
 
 	// this is only for virtio-blk and virtio-scsi support
 	GetAndSetSandboxBlockIndex() (int, error)
 	DecrementSandboxBlockIndex() error
 
-	// this is for vhost_user devices
-	AddVhostUserDevice(VhostUserDevice, config.DeviceType) error
-}
-
-// VhostUserDevice represents a vhost-user device. Shared
-// attributes of a vhost-user device can be retrieved using
-// the Attrs() method. Unique data can be obtained by casting
-// the object to the proper type.
-type VhostUserDevice interface {
-	Attrs() *config.VhostUserDeviceAttrs
-	Type() config.DeviceType
+	// this is for appending device to hypervisor boot params
+	AppendDevice(Device) error
 }
 
 // Device is the virtcontainers device interface.
 type Device interface {
 	Attach(DeviceReceiver) error
 	Detach(DeviceReceiver) error
+	// ID returns device identifier
+	DeviceID() string
+	// DeviceType indicates which kind of device it is
+	// e.g. block, vfio or vhost user
 	DeviceType() config.DeviceType
+	// GetDeviceInfo returns device specific data used for hotplugging by hypervisor
+	// Caller could cast the return value to device specific struct
+	// e.g. Block device returns *config.BlockDrive and
+	// vfio device returns []*config.VFIODev
+	GetDeviceInfo() interface{}
+	// IsAttached checks if the device is attached
+	IsAttached() bool
 }
 
 // DeviceManager can be used to create a new device, this can be used as single
 // device management object.
 type DeviceManager interface {
```

Comment:
@WeiZhang555 I'm wondering how this can be consumed by frakti. Currently sandbox has its private device manager and that will manage all device hotplug internally. If frakti wants to use this API, it will have to create a device manager on its own -- so that it can call functions in the DeviceManager interface. But that seems to conflict with the internal device manager in the sandbox structure. Would you please add device manager as an argument to `CreateSandbox`/`RunSandbox` and `FetchSandbox`, and use that to set sandbox.devManager if it's non-nil?

Another way is to make `sandbox.HotplugAddDevice`/`sandbox.HotplugRemoveDevice` top level API and use device manager underneath. But since you have already chosen a different path, there is no need to go back there. Thanks for the hard work!

Reply:
@bergwolf 

> Another way is to make sandbox.HotplugAddDevice/sandbox.HotplugRemoveDevice top level API and use device manager underneath. 

This is what I intended to do in seperate PR , and because this PR is too huge, my initial plan was only keeping the rework part in this one.

What do you think of the `sandbox.HotplugAddDevice/sandbox.HotplugRemoveDevice` ?

Reply:
And `sandbox.HotplugAddDevice` will be simply wrapper of `DeviceManager.HotplugAddDevice`

Reply:
Sounds good to me. Thanks!